### PR TITLE
YLP-2277 Convert glycemia values into user defined unit (mmol/L <-> mg/dL)

### PR DESCRIPTION
**Pull request by yanngit**

### CODE REVIEW

**Review by CedricPerotto — CHANGES_REQUESTED** [review]

File: usecase/helper.go [hunk 1]
```diff
@@ -252,14 +256,34 @@ func (p *PatientData) writeDataToBuffer(
...
+				return fmt.Errorf("cannot parse startDate=%s", dates.Start)
+			}
+		}
+		if dates.End != "" {
+			if endDate, err = time.Parse(time.RFC3339Nano, dates.End); err != nil {
+				return fmt.Errorf("cannot parse endDate=%s", dates.End)
+			}
+		} else {
+			endDate = time.Now()
+		}
```
[the remark this answers is not in the document]
Comment (yanngit) [reply usecase/helper.go]: Fixed

File: Jenkinsfile [hunk 2]
```diff
@@ -72,7 +72,7 @@ pipeline {
             }
         }
         stage('Publish') {
-            when { branch "dblp" }
+//             when { branch "dblp" }
```
Comment (ydrisR) [inline Jenkinsfile:75]: Uncomment and change dblp -> main

**Review by ydrisR — CHANGES_REQUESTED** [review]

File: usecase/helper.go [hunk 3]
```diff
@@ -274,10 +291,40 @@ func writeDeviceParameterChanges(res *bytes.Buffer, p *writeFromIter) error {
...
 		datum["value"] = paramChange.Value
 		datum["level"] = paramChange.Level
+
 		if paramChange.PreviousValue != "" {
 			datum["previousValue"] = paramChange.PreviousValue
 		}
 
+		/*Handle conversion if bgUnit is not empty*/
+		if bgUnit != "" {
+			if datum["units"] != bgUnit {
```
Comment (CedricPerotto) [inline usecase/helper.go:301]: we should also check that datum["units"] is MgdL or MmolL

File: usecase/helper.go [hunk 4]
```diff
@@ -274,10 +291,40 @@ func writeDeviceParameterChanges(res *bytes.Buffer, p *writeFromIter) error {
...
+				if datum["units"] == MgdL {
+					datum["units"] = MmolL
+					datum["value"] = fmt.Sprintf("%g", getMmol(val))
+				} else if datum["units"] == MmolL {
+					datum["units"] = MgdL
+					datum["value"] = fmt.Sprintf("%g", getMgdl(val))
+				}
+			}
+
+			if paramChange.PreviousUnit != bgUnit && paramChange.PreviousValue != "" {
```
Comment (CedricPerotto) [inline usecase/helper.go:315]: same here we should also check that paramChange.PreviousUnit is MgdL or MmolL

File: usecase/helper.go [hunk 5]
```diff
@@ -312,14 +367,64 @@ func writePumpSettings(res *bytes.Buffer, p *writeFromIter) error {
 	/*TODO fetch from somewhere*/
 	datum["activeSchedule"] = "Normal"
 	datum["deviceId"] = settings.CurrentSettings.Device.DeviceID
+
+	/* Perform conversion */
+	if bgUnit != "" {
+		for _, hp := range settings.HistoryParameters {
+			if hp.Unit != bgUnit {
```
Comment (CedricPerotto) [inline usecase/helper.go:374]: same here we should also check that hp.Unit is MgdL or MmolL

File: usecase/helper.go [hunk 6]
```diff
@@ -312,14 +367,64 @@ func writePumpSettings(res *bytes.Buffer, p *writeFromIter) error {
...
+				if hp.Unit == MgdL {
+					hp.Unit = MmolL
+					hp.Value = fmt.Sprintf("%g", getMmol(val))
+				} else if hp.Unit == MmolL {
+					hp.Unit = MgdL
+					hp.Value = fmt.Sprintf("%g", getMgdl(val))
+				}
+			}
+
+			if hp.PreviousUnit != bgUnit {
```
Comment (CedricPerotto) [inline usecase/helper.go:388]: same here we should also check that hp.PreviousUnit is MgdL or MmolL

File: usecase/helper.go [hunk 7]
```diff
@@ -374,8 +479,16 @@ func groupByChangeDate(parameters []orcaSchema.HistoryParameter) []GroupedHistor
...
 }
 
+func getMgdl(value float64) float64 {
+	return math.Round(value * MmolLToMgdLConversionFactor)
+}
+
+func getMmol(value float64) float64 {
+	return math.Round(value / MmolLToMgdLConversionFactor)
+}
+
```
Comment (CedricPerotto) [inline usecase/helper.go:489]: tide-v2 has this implementation in order to limit the number of digits in the float part:

```go
func convertToMmoll(value float64) float64 {
	roundedValue := math.Round(value / schema.MmolLToMgdLConversionFactor * schema.MmolLToMgdLPrecisionFactor)
	floatValue := roundedValue / schema.MmolLToMgdLPrecisionFactor
	return floatValue
}
```

File: usecase/helper_test.go [hunk 8]
```diff
@@ -164,9 +152,37 @@ func TestAPI_getLatestPumpSettings_handleNotFound(t *testing.T) {
...
+
+	/*When*/
 	res, err := usecase.getLatestPumpSettings(timeContext, "traceId", userId, &writer, token)
+
+	/*Then*/
 	assert.Nil(t, err)
 	assert.Nil(t, res)
 }
+
+func Test_getMgdl(t *testing.T) {
```
Comment (CedricPerotto) [inline usecase/helper_test.go:165]: No tests added for getMmol ?

File: usecase/patientData.go [hunk 9]
```diff
@@ -75,13 +83,13 @@ type (
 		GlyHypoLimit float64 `json:"glyHypoLimit"`
 		// The Hyper limit used to compute TIR & TBR
 		GlyHyperLimit float64 `json:"glyHyperLimit"`
-		// The unit of hypo/hyper values
+		// The given of hypo/hyper values
```
Comment (CedricPerotto) [inline usecase/patientData.go:86]: Is it a replace all error ?

File: usecase/patientData.go [hunk 10]
```diff
@@ -75,13 +83,13 @@ type (
...
 		GlyUnit string `json:"glyUnit"`
 	}
 	deviceParameter struct {
 		Level int    `json:"level" bson:"level"`
 		Name  string `json:"name" bson:"name"`
-		Unit  string `json:"unit" bson:"unit"`
+		Unit  string `json:"given" bson:"given"`
```
Comment (CedricPerotto) [inline usecase/patientData.go:92]: Same here, I don't get why this field changes from unit to given...

File: api/patientDataController.go [hunk 11]
```diff
@@ -36,11 +36,27 @@ func (a *API) getDataV2(ctx context.Context, res *common.HttpResponseWriter) err
 	endDate := query.Get("endDate")
 	withPumpSettings := query.Get("withPumpSettings") == "true"
 	sessionToken := getSessionToken(res)
-	err := a.patientData.GetData(ctx, userID, res.TraceID, startDate, endDate, withPumpSettings, sessionToken, &buffer)
+	bgUnit := query.Get("bgUnit")
+	if bgUnit != usecase.MgdL && bgUnit != usecase.MmolL {
+		bgUnit = ""
+	}
+	getDataArgs := usecase.GetDataArgs{
+		Ctx:                        ctx,
```
Comment (CedricPerotto) [inline api/patientDataController.go:44]: passing context in structs is not recommended 
https://go.dev/blog/context-and-structs#Storing%20Context%20in%20Structs%20Leads%20to%20Confusion

File: usecase/patientData.go [hunk 12]
```diff
@@ -434,17 +434,17 @@ func writeFromIterV1(ctx context.Context, res *bytes.Buffer, bgUnit string, p *w
...
+							datum["value"] = convertToMgdl(datum["value"].(float64))
 						} else {
 							datum["units"] = MmolL
-							datum["value"] = getMmol(datum["value"].(float64))
+							datum["value"] = convertToMmol(datum["value"].(float64))
 						}
 					}
 				case "wizard":
-					/*For wizard, we don't have anymore fields in mmol, so we're changing the unit but no conversion is done.
+					/*For wizard, we don't have anymore fields in mmol, so we're changing the given but no conversion is done.
```
Comment (CedricPerotto) [inline usecase/patientData.go:447]: less impact but looks like the unit -> given replace may be not appropriate here

**Review by CedricPerotto — CHANGES_REQUESTED** [review]

**Pull request merged**